scylla/src/transport/locator/tablets.rs

```diff
@@ -95,44 +96,120 @@ pub(crate) struct Tablet {
     first_token: Token,
     last_token: Token,
     replicas: TabletReplicas,
+    /// If any of the replicas failed to resolve to a node,
+    /// then this field will contain the original list of replicas.
+    failed: Option<RawTabletReplicas>,
 }
 
 #[derive(Clone, Debug, Default)]
 pub(crate) struct TableTablets {
     tablet_list: Vec<Tablet>,
+    /// In order to make typical tablet maintance faster
+    /// we remember if there were any tablets that have unrecognized uuids in replica list.
+    /// If there were none, and few other conditions are satisfied, we can skip nearly whole maintanace.
+    /// This flag may be falsely true: if we add tablet with unknown replica but later
+    /// overwrite it with some other tablet.
+    has_unknown_replicas: bool,
 }
 
 #[derive(Clone, Debug)]
 pub(crate) struct TabletsInfo {
     tablets: HashMap<TableSpec, TableTablets>,
+    /// See `has_unknown_replicas` field in `TableTablets`.
+    /// The field here will be true if it is true for any TableTablets.
+    has_unknown_replicas: bool,
 }
 
 impl Tablet {
     pub(crate) fn from_raw_tablet(
-        raw_tablet: &RawTablet,
+        raw_tablet: RawTablet,
         replica_translator: impl Fn(Uuid) -> Option<Arc<Node>>,
     ) -> Self {
+        let replicas_result =
+            TabletReplicas::from_raw_replicas(&raw_tablet.replicas, replica_translator);
+        let (replicas, failed) = match replicas_result {
+            Ok(r) => (r, None),
+            Err((r, f)) => {
+                info!("Nodes ({}) from system.tablets not present in current ClusterData.known_peers. \
+                       Skipping this replica until topology refresh", f.iter().format(", "));
+                (r, Some(raw_tablet.replicas))
+            }
+        };
         Self {
             first_token: raw_tablet.first_token,
             last_token: raw_tablet.last_token,
-            replicas: TabletReplicas::from_raw_replicas(&raw_tablet.replicas, replica_translator),
+            replicas,
+            failed,
+        }
+    }
+
+    // Returns true if after the operation Tablet replicas are fully resolved
+    fn re_resolve_replicas(
+        &mut self,
+        replica_translator: impl Fn(Uuid) -> Option<Arc<Node>>,
+    ) -> bool {
+        if let Some(failed) = self.failed.as_ref() {
+            match TabletReplicas::from_raw_replicas(failed, replica_translator) {
+                Ok(resolved_replicas) => {
+                    // We managed to successfully resolve all replicas, all is well.
+                    self.replicas = resolved_replicas;
+                    self.failed = None;
+                    true
+                }
+                Err((_, failed)) => {
+                    warn!("Nodes ({}) listed as replicas for a tablet are not present in ClusterData.known_peers, \
+                           despite topology refresh. Removing problematic tablet.", failed.iter().format(", "));
```

Comment:
How? Print it's token range?

Reply:
If the token range serves as the best identifier of a tablet, then yes. Anything that could come in handy while debugging will be appreciated.

Reply:
The token range alone is not precise - keyspace name and table name are missing, but we don't have access to them in `Tablet`

Reply:
Token range is not precise. `Table` method that this log is located in could return Result with list of failed replicas instead of bool. Then I'd have to print the token range and table name in `TableTablets` - which doesn't store table name, but I can add it - it's one allocation per table, so it's ok

Reply:
Yes, one allocation per table seems acceptable.

Reply:
Moved the error message to `TableTablets::perform_maintanance`, with precise tablet identification

Reply:
I also moved the message from `Tablet::from_raw_tablet` to cluster.rs and improved in in the same way (printing precise tablet info)